File: 0987-vertical-order-traversal-of-a-binary-tree/0987-vertical-order-traversal-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
         vector<tuple<int, int, int>> nodes;

        function<void(TreeNode*, int, int)> dfs = [&](TreeNode* node, int x, int y) {
            if (!node) return;
            nodes.emplace_back(x, y, node->val);
            dfs(node->left, x - 1, y + 1);
            dfs(node->right, x + 1, y + 1);
        };

        dfs(root, 0, 0);

        sort(nodes.begin(), nodes.end());

        vector<vector<int>> res;
        int prev_x = INT_MIN;
        for (auto& [x, y, val] : nodes) {
            if (x != prev_x) {
                res.emplace_back();
                prev_x = x;
            }
            res.back().push_back(val);
        }

        return res;
    }
};

File: 0987-vertical-order-traversal-of-a-binary-tree/0987-vertical-order-traversal-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        nodes = []
# brute force
        def dfs(node, x, y):
            if not node:
                return
            nodes.append((x, y, node.val))
            dfs(node.left, x - 1, y + 1)
            dfs(node.right, x + 1, y + 1)

        dfs(root, 0, 0)

        nodes.sort()  # sorts by x, then y, then value

        res = []
        prev_x = float('-inf')
        for x, y, val in nodes:
            if x != prev_x:
                res.append([])
                prev_x = x
            res[-1].append(val)

        return res